system/lib/pthread/proxying.c

```diff
@@ -330,24 +335,26 @@ void emscripten_proxy_execute_queue(em_proxying_queue* q) {
 
   pthread_mutex_lock(&q->mutex);
   task_queue* tasks = get_tasks_for_thread(q, pthread_self());
-  if (tasks == NULL || tasks->processing) {
-    // No tasks for this thread or they are already being processed.
-    goto end;
-  }
-  // Found the task queue; process the tasks.
-  tasks->processing = 1;
-  while (!task_queue_is_empty(tasks)) {
-    task t = task_queue_dequeue(tasks);
-    // Unlock while the task is running to allow more work to be queued in
-    // parallel.
-    pthread_mutex_unlock(&q->mutex);
-    t.func(t.arg);
-    pthread_mutex_lock(&q->mutex);
+  pthread_mutex_unlock(&q->mutex);
+
+  if (tasks != NULL) {
+    pthread_mutex_lock(&tasks->mutex);
+    if (!tasks->processing) {
```

Comment:
Can we move this check up two lines and make it `if (tasks && !tasks->processing)`..  no other thread can process the queue so there should be no races around `tasks->processing`.

Reply:
I think we can then avoid the inner unlock around t.func()?

Reply:
We don't want to keep the lock while executing the task in case someone tries to queue more work at the same time, but you're right that `processing` is only ever accessed on the current thread so we don't need the lock to access it. Will simplify the code.